### ISSUE TITLE
Environment variables are not always set when SSHing into container (from the SSHD script)

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-dev-containers#683]
SSH creates a "login" shell by default (e.g. `bash -li`).  The `-l` flag results in a fresh login shell which pulls environment variables out of /etc/environment, then fires ~/.profile (and by extension ~/.bashrc).

The problem is that the `ENV` directive and `-e` inject environment variables into the container, so there's nothing in classic locations like `/etc/environment` to populate a fresh shell. For example, consider the `julia` image from DockerHub:

```bash
# Interactive shell
$ docker run -it --rm julia /bin/bash -ic 'echo $PATH'
/usr/local/julia/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

# Login, interactive shell
$ docker run -it --rm julia /bin/bash -lic 'echo $PATH'
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

So an interactive shell is safe for largely any image, but login is not. 

The SSHD script in the script library does not solve this problem, so any dev container like codespaces-linux that uses it would see the same thing unless the PATH and other env vars are exxplicitly set in `/etc/profile`, ` /etc/bashrc.bash` or their user specific equivalents.

We need to determine if there is an easy way to fix this in config.

**Comment by Chuxel** [comment]
It looks like we should be able to have the script create `/etc/profile.d/restore-env.sh` to fix this using the following syntax:

```bash
echo "export PATH=\${PATH}:${PATH//$(bash -lc 'echo $PATH')/}" > /etc/profile.d/restore-env.sh
```

This will get the login path, remove it from the current path, then add the delta into an export statement in restore-env.sh.  This is then automatically sourced on login.

**Comment by chrmarti** [comment]
The pattern I have seen is that some `/etc/profile`s replace `PATH` completely. I'm not sure if the image's `PATH` will always contain the login shell's `PATH` as a substring like your suggestion assumes.

I like the idea of fixing this in `/etc/profile.d/...` as this would also allow us to use an interactive login shell for probing the environment in Remote-Containers.

**Comment by Chuxel** [comment]
@chrmarti I'm queuing up a PR to fix this for our own definitions in places that where the environment is updated and therefore the use of `ENV` is also likely using the filename `/etc/profile.d/00-remote-env.sh`.

**Comment by Chuxel** [comment]
@chrmarti Should we open another one for doing this automatically for images that are not out of this repo?

**Comment by chrmarti** [comment]
@Chuxel Should we consider the case where `/etc/profile` sets a `PATH` that is not a substring of the image's `PATH`?

**Comment by Chuxel** [comment]
@chrmarti For the generic case, yes, that's a good point.  That said, a bit unclear what we'd want to do in that case.  What would happen here, is it would just overwrite the PATH with the same values as in the non-login case.  The only alternative I can think of is to add `${PATH}` to either the beginning or end of the string to get the union of the two in this scenario.  Thoughts?

**Comment by chrmarti** [comment]
Would the following work?: We could go through the image's path entries and check if it is already part of the current PATH and if not we insert it after the last entry in the current PATH that is also in the image's path.

In TypeScript (haven't tried yet in /bin/sh):
```ts
function mergePaths(currentPath: string, imagePath: string) {
	const result = currentPath.split(':');
	let insertAt = 0;
	for (const entry of imagePath.split(':')) {
		const i = result.indexOf(entry);
		if (i === -1) {
			result.splice(insertAt++, 0, entry);
		} else {
			insertAt = i + 1;
		}
	}
	return result.join(':');
}
```

One slightly off case is when the user's PATH does not have the `sbin` entries, but the image's path has them, they are then added back in. That shouldn't be a big problem though since we currently run the regular users with the `sbin` entries from the image by default.

**Comment by chrmarti** [comment]
We could hardcode `sbin` as a special-case or use a variation of your diff approach by not adding those entries that are missing, but are present in root's PATH:
```ts
function mergePaths(currentPath: string, imagePath: string, rootPath: string) {
	const result = currentPath.split(':');
	const rootEntries = rootPath.split(':');
	let insertAt = 0;
	for (const entry of imagePath.split(':')) {
		const i = result.indexOf(entry);
		if (i === -1) {
			if (rootEntries.indexOf(entry) === -1) {
				result.splice(insertAt++, 0, entry);
			}
		} else {
			insertAt = i + 1;
		}
	}
	return result.join(':');
}
```

**Comment by Chuxel** [comment]
> One slightly off case is when the user's PATH does not have the sbin entries, but the image's path has them, they are then added back in. That shouldn't be a big problem though since we currently run the regular users with the sbin entries from the image by default.

@chrmarti Yeah the interesting thing is that is the default for non-root users.  Here's the snippet in `/etc/profile`:

```bash
if [ "`id -u`" -eq 0 ]; then
  PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
else
  PATH="/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games"
fi
export PATH
```

It just so happens that our containers generally start as root, so they get the `sbin` path in the Dockerfile. When you do this in the container itself and running as a non-root user, you get the other path.

```bash
$ sudo su vscode
$ bash -lc 'echo $PATH'
/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
$ sudo su
$ bash -lc 'echo $PATH'
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

Generally `sbin` points to `bin` but does allow you to add root-only commands, which is why it doesn't show up in the other case.

**Comment by Chuxel** [comment]
@chrmarti @bamurtaugh Should we move this to vscode-remote-release for having Remote - Containers do this automatically, or just close it since the scripts in this repo address it?

**Comment by bamurtaugh** [comment]
@Chuxel I don't think I've seen this highly discussed in vscode-remote-release. However, since something like https://github.com/microsoft/vscode-remote-release/issues/2994 could bring this top of mind if my understanding is correct (unless that would be a separate scenario/implementation that wouldn't be impacted by this), perhaps it'd make sense to move this issue to remote-release and address it once we work on https://github.com/microsoft/vscode-remote-release/issues/2994?

**Comment by Chuxel** [comment]
@bamurtaugh Yeah this is a bit confusing... the reverse is happening here. :) In this case, the container has a SSH server in it and you connect to it (rather than wanting to connect to a SSH server to host containers).

The real issue it turned out was not that, and how login shells (`bash -l`) were wiping out anything passed in by Docker. I adapted the library scripts where this could happen to drop in a script into `profile.d` to resolve it, but it could be generalized into the dev container startup logic.

**Comment by chrmarti** [comment]
When the image's `PATH` is workable, it is enough to change `/etc/profile` to not overwrite `PATH`. E.g. using `sed` in the Dockerfile:
```Dockerfile
RUN sed -i 's/PATH=/# PATH=/g' /etc/profile
```

Customizations in `/etc/profile.d` would still be applied.

As a side note: There is `userEnvProbe: 'none' | 'loginInteractiveShell' | 'interactiveShell' | 'loginShell'` in the devcontainer.json to customize the VS Code Server's environment, but that does not apply to SSH logins.

(Moving to vscode-remote-release.)
